src/Perl6/World.nqp

```diff
@@ -530,44 +530,72 @@ class Perl6::World is HLL::World {
           $want
         ) == -1
     }
+
+    method !check-for-PREVIEW ($ver-match, $rev, $default_rev, $rev_mod) {
+        $ver-match.PRECURSOR.worry('PREVIEW modificator is used with released specification 6.' ~ $rev)
+            if nqp::isle_s($rev, $default_rev) && nqp::iseq_s($rev_mod, 'PREVIEW');
+
+        $ver-match.PRECURSOR.worry(
+            'Language specification 6.' ~ $rev ~ ' is not released yet, PREVIEW modificator is expected'
+        ) if nqp::isgt_s($rev, $default_rev) && nqp::isne_s($rev_mod, 'PREVIEW');
+    }
+
     method load-lang-ver($ver-match, $comp) {
         $*MAIN   := 'MAIN';
         $*STRICT := 1 if $*begin_compunit;
 
         my str $version := ~$ver-match;
-        # fast-path the common cases
-        if $version eq 'v6.c' {
-            $comp.set_language_version: '6.c';
-            $*CAN_LOWER_TOPIC := 0;
-            # CORE.c is currently our lowest core, which we don't "load"
-            return;
-        }
+        my @vparts := nqp::split('.', $version);
+        my $default_rev := nqp::substr($comp.config<language-version>, 2, 1);
 
-        if $version eq 'v6' ?? nqp::substr($comp.language_version, 2, 1)
-        !! $version eq 'v6.d' ?? 'd'
-        !! $version eq 'v6.d.PREVIEW' ?? 'd'
-        !! '' -> $lang {
-            $comp.set_language_version:       '6.' ~ $lang;
-            self.load_setting: $ver-match, 'CORE.' ~ $lang;
-            return;
+        # Do we have dot-splitted version string?
+        if (@vparts > 1) || ($version eq 'v6') {
+            my $revision := @vparts[1] || $default_rev;
+            my $lang_ver := '6.' ~ $revision;
+
+            self."!check-for-PREVIEW"($ver-match, $revision, $default_rev, @vparts[2] || '')
+                if ($revision eq 'c') || ($lang_ver eq $comp.config<language-version>);
```

Comment:
Hm, why the explicit check for `'c'` here? Should we not be getting the language version (`c`, `d`, etc.) and comparing it with `le` against the version we're currently at?

Reply:
There is no `CORE.c.setting`, thus – special case for `c`.